### ISSUE TITLE
Save your answers suggesting you need to make an account

### ISSUE THREAD

**Issue by BryceStevenWilley** [mplp/docassemble-Simpmotion#17]
The `Navigation and Saving Your Form` page says:

> If you can’t finish today, you can save your answers by creating a username and password. This lets you log back into the site later and finish, or go back and change any of your answers.

While this works, users don't have to create an account; their answers will be saved, even if they close out the tab. They'd have to have the correct link to get back to the same interview, but if you visit something like https://da-dev.mplp.org/run/Simpmotion/simpmotion and haven't cleared your browser's cookies or gone past the 60 days that DA will auto-delete interviews, you can continue as normal.

It's definitely a bit complicated to explain all that though. I'd consider maybe saying "Create a username and password, or bookmark this page to continue where you left off". 

(Realized the best way to review existing code was just to make individual issues about stuff I wasn't sure how to change, and I'll make a PR with changes that I'm more confident about).

**Comment by scoellis** [comment]
Hey Bryce, do you know if anyone is successfully using Google, Facebook,
Apple, Microsoft, or any of the common Oauth providers to allow users to
create accounts and save their answers? I looked into creating an Oauth app
in our Google Workspace domain, but we decided to wait and talk to others
about it before going down that path.

---
Scott Ellis
(he/him)
IT Systems Administrator
Michigan Poverty Law Program
(734) 714-3234
***@***.***


*This electronic communication may be subject to the attorney-client
privilege and may contain confidential information. If you are not the
intended recipient, any distribution, copying or disclosure is strictly
prohibited.  If you have received this communication in error, please
notify the sender immediately and delete this copy from your system. Thank
you for your cooperation.*

**Comment by BryceStevenWilley** [comment]
I don't know of anyone I've worked with before using it, but it is in docassemble, and is something Jonathan is actively maintaining. It's something I'll likely be working on / implementing for Massachusetts next year (our e-filing process will need to be more tightly integrated to OAuth at that point). Totally understand if we don't want to be the first to try it out though.

https://docassemble.org/docs/config.html#oauth is the docassemble documentation (in case you haven't seen it yet). I think the most complicated part would be generating the OAuth secret in Google, if you wanted to do that, we could integrate it in the dev server, and it'd be pretty clear if it works well or not.

**Comment by scoellis** [comment]
I think I can generate the Oauth secret in Google. I'll work on that and
put in Bitwarden and let you all know when I add it.

---
Scott Ellis
(he/him)
IT Systems Administrator
Michigan Poverty Law Program
(734) 714-3234
***@***.***


*This electronic communication may be subject to the attorney-client
privilege and may contain confidential information. If you are not the
intended recipient, any distribution, copying or disclosure is strictly
prohibited.  If you have received this communication in error, please
notify the sender immediately and delete this copy from your system. Thank
you for your cooperation.*

**Comment by scoellis** [comment]
Okay, I created an Oauth secret and added it to the Bitwarden docassemble
development server item.

I'm not sure how well it will work but I'd like to give it a try. As of
right now I only have it set up to use with one test account (my personal
account).

I don't mind adding the information to the Config YML file, unless we want
one person to be making changes to that file since it does restart the
server.

Thanks,
Scott

---
Scott Ellis
(he/him)
IT Systems Administrator
Michigan Poverty Law Program
(734) 714-3234
***@***.***


*This electronic communication may be subject to the attorney-client
privilege and may contain confidential information. If you are not the
intended recipient, any distribution, copying or disclosure is strictly
prohibited.  If you have received this communication in error, please
notify the sender immediately and delete this copy from your system. Thank
you for your cooperation.*

**Comment by ekressmiller** [comment]
I don't have strong feelings about only one person changing the config file, but I'm also happy to do it if you want.

**Comment by scoellis** [comment]
Okay, I'll go ahead and try it out one evening and see if it works. If it
doesn't work I'll just revert the Config file. Thanks.

---
Scott Ellis
(he/him)
IT Systems Administrator
Michigan Poverty Law Program
(734) 714-3234
***@***.***


*This electronic communication may be subject to the attorney-client
privilege and may contain confidential information. If you are not the
intended recipient, any distribution, copying or disclosure is strictly
prohibited.  If you have received this communication in error, please
notify the sender immediately and delete this copy from your system. Thank
you for your cooperation.*

**Comment by ekressmiller** [comment]
Returning to @BryceStevenWilley's original point in this thread, my instinct is that we should err on the side of telling people to make an account. If they're in a public library or if someone else uses their computer and clears the cache or something, the session would be lost, right? I wouldn't want people to depend on it still being there. 

But in our instructions article and when dealing with users on livehelp, if someone is trying to get back to saved answers, we could tell them that just returning to the interview might get them back in. 

Does that make sense to others?

**Comment by BryceStevenWilley** [comment]
You're right that making an account does make things easier for users, but it is a fairly large ask that will turn a lot of people away. Looking back at the original wording, I think that my issue is with the implication that their answers will be lost if they can't finish it within a certain time period.

> If you can’t finish today, you can save your answers by creating a username and password.

IMO something like this would be better, and still suggest people make an account while being a bit more precise about when they might need to.

> If you need to close this window before you finish, you can save your answers by creating a username and password.

(depending on browser settings, they can sometimes clear all cookies when the browser is closed, so our suggestions should probably be meant for the strictest settings)

---

> if someone is trying to get back to saved answers, we could tell them that just returning to the interview might get them back in.

That is right, links like https://apps-dev.suffolklitlab.org/run/ThePackageName/the_yaml_name (that have "run" in them) will get you back to the interview's most current session with the current user (either from the logged in user or from the cache).  Links with "start" instead of "run" will start the interview over from the beginning, and I think if the user isn't logged in, they won't be able to access that older session (another point towards encouraging people to login).

**Comment by ekressmiller** [comment]
Yeah, I like your language edit to the first part @BryceStevenWilley. @normon66 tagging you too in case you have thoughts.

To increase the chances that people can get back into an interview they started if they didn't create an account, I guess we would want all the links from Michigan Legal Help to be /run/ links instead of /start/ links.  That way if the same user is coming back, they should return to the same point in their interview, but if it's a new user, it will default to the start page.

There's a chance for confusion for some people with shared computers or if they opened the interview in the past, maybe fiddled around with it, but are now expecting to start it over for real. Also could be an increased safety risk for PPO-type cases. (If abuser opens /run/ version of interview and sees survivor's answers there.)  Maybe we'd want to consider using /start/ links or adding some specific warnings for those.  

This is probably something to run by at least Ang and maybe more team members too.

**Comment by BryceStevenWilley** [comment]
> There's a chance for confusion for some people with shared computers or if they opened the interview in the past, maybe fiddled around with it, but are now expecting to start it over for real. Also could be an increased safety risk for PPO-type cases. (If abuser opens /run/ version of interview and sees survivor's answers there.) Maybe we'd want to consider using /start/ links or adding some specific warnings for those.

Good points. For safety risk, there would be other ways to access the answers, like going through someone's browser history, as the links there are `/run/` links. I feel like the chance that an abuser navigates to the MLH homepage / launch page but not directly to the interview itself (or that a survivor would delete their history of the interview but not the landing page) seems very low, but I'm not 100% sure.

I definitely agree that it should be a discussion with more people; unfortunately a lot of the changes necessary for safety do result in a worse user-experience, so we do need to decide which to prioritize.

**Comment by normon66** [comment]
Thanks for looping me in! I had not thought about this yet, specifically in regards to how docassemble saves progress, regardless.

If we want users to be able to start from where they left off, I think it's necessary that they should only do so with saved answers in their own account. They can still use it as a guest (as with LHI), but should have an account to save anything. 

Whether a protection order, a fee waiver, or just a simple letter, we shouldn't have users' information out and available for the next person who happens to use the same interview with the `/run/` in it. Similarly, since multiple users might use the same library computer or self-help kiosk to create forms, I think we should always go with the `/start/` option. If user A gets part way through an interview and user B picks up the same interview later, even if they don't see any of user A's personal information, they still might miss other important information that was in previous screens. 

As with many of our discussions, I'm certain we're not the first group to discuss this! Along with looping in Ang/MLH members, this might be useful to put into Teams next week or bring up on Monday.

**Comment by ekressmiller** [comment]
Agree this could be a good one to discuss on the Monday calls. It slipped my mind yesterday, but maybe next week!
